cylc/flow/param_expand.py

```diff
@@ -201,6 +201,73 @@ def _expand_name(self, results, tmpl, params, spec_vals=None):
                 spec_vals[params[0][0]] = param_val
                 self._expand_name(results, tmpl, params[1:], spec_vals)
 
+    @staticmethod
+    def _parse_parent_string(parent):
+        """Takes a parent string and returns a list of parameters and a
+        template string.
+
+        Examples:
+            >>> this = NameExpander._parse_parent_string
+
+            # Parent doesn't contain a parameter:
+            >>> this('foo')
+            ([], 'foo')
+
+            # Parent contains a simple single parameter:
+            >>> this('<foo>')
+            (['foo'], '{foo}')
+
+            # Parent contains 2 parameters in 1 <>:
+            >>> this('something<foo, bar>other')
+            (['foo', 'bar'], 'something{foo}{bar}other')
+
+            # Parent contains 2 parameters in 2 <>:
+            >>> this('something<foo>middlebit<bar>other')
+            (['foo', 'bar'], 'something{foo}middlebit{bar}other')
+
+            # Parent contains 2 parameters, once with an = sign in it.
+            >>> this('something<foo=42>middlebit<bar>other')
+            (['foo=42', 'bar'], 'something{foo}middlebit{bar}other')
+
+            # Parent contains 2 parameters in 2 <>:
+            >>> this('something<foo,bar=99>other')
+            (['foo', 'bar=99'], 'something{foo}{bar}other')
+
+            # Parent contains spaces around = sign:
+            >>> this('FAM<i = cat ,j=3>')
+            (['i = cat', 'j=3'], 'FAM{i}{j}')
+        """
+        tmpl_list = re.split(r'(<[^>]*>)', parent)
+        raw_param_list = [
+            REC_P_GROUP.findall(i)[0]
+            for i in tmpl_list
+            if REC_P_GROUP.findall(i)
+        ]
+        param_list = []
+        for param in raw_param_list:
+            if ',' in param:
+                # parameter syntax `<foo, bar>`
+                sub_params = [
+                    i.strip() for i in param.split(',')]
+                for sub_param in sub_params:
+                    param_list.append(sub_param)
+                    if '=' in sub_param:
+                        sub_params[
+                            sub_params.index(sub_param)
+                        ] = sub_param.split('=')[0].strip()
+                replacement = '{' + '}{'.join(sub_params) + '}'
```

Comment:
This logic must already be happening somewhere else in the code so is presumably happening twice with this addition?

Looks like it might be covered by `REC_P_OFFS`?

Reply:
It's happening in the `expand` method. Unfortunately I'm struggling to get the interface to work elegantly - I think it might be worth thoroughly testing and a refactor at some point. I strongly suspect  that `tests/functional/param-expand` is a prime candidate for replacement (or at least pruning) with unit tests and functional tests..